pkg/rbd/controllerserver.go

```diff
@@ -230,18 +244,43 @@ func (cs *ControllerServer) checkSnapshot(ctx context.Context, req *csi.CreateVo
 	defer cr.DeleteCredentials()
 
 	rbdSnap := &rbdSnapshot{}
-	if err = genSnapFromSnapID(ctx, rbdSnap, snapshotID, cr); err != nil {
-		if _, ok := err.(ErrSnapNotFound); !ok {
+	if acquired := cs.SnapshotLocks.TryAcquire(snapshotID); !acquired {
+		klog.Infof(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), snapshotID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, snapshotID)
+	}
+	defer cs.SnapshotLocks.Release(snapshotID)
+
```

Comment:
Need to enforce flattening limits prior to the creation of the RBD snapshot below since you cannot flatten after the snapshot it created (i.e. it might already be running the background=).